### ISSUE TITLE
Expose GetInvocationList from F# event

### ISSUE THREAD

**Issue by xperiandri** [fsharp/fslang-suggestions#938]
# Title of Suggestion

I have an expensive logic that I want to execute only if there are any subscribers to the event. Like this

``` F#
  let modelTypeChanged = Event<EventHandler, EventArgs>()
  [<CLIEvent>]
  member val ModelTypeChanged = modelTypeChanged.Publish
  member vm.CurrentModel
    with get () : 'model = currentModel
    and private set (newModel  : 'model) =
        // TODO:
        //if currentModel <> newModel then
            let oldModel = currentModel
            currentModel <- newModel
            notifyPropertyChanged "CurrentModel"
            let ``delegate`` = modelTypeChanged |> getDelegateFromEvent
            match ``delegate`` with
            | null -> ()
            | _ when ``delegate``.GetInvocationList () |> (not << Seq.isEmpty) ->
                let oldModelType = oldModel.GetType ()
                let newModelType = newModel.GetType ()
                match FSharpType.IsUnion newModelType, FSharpType.IsUnion newModelType with
                | true, true ->
                    let oldUnionCase, _ = FSharpValue.GetUnionFields(oldModel, oldModelType)
                    let newUnionCase, _ = FSharpValue.GetUnionFields(newModel, newModelType)
                    if oldUnionCase.DeclaringType <> newUnionCase.DeclaringType
                    || oldUnionCase.Name <> newUnionCase.Name
                    then modelTypeChanged.Trigger (vm, EventArgs ())
                | false, false ->
                    if oldModelType <> newModelType
                    then modelTypeChanged.Trigger (vm, EventArgs ())
                | _ -> modelTypeChanged.Trigger (vm, EventArgs ())
            | _ -> ()
        //else ()
```

But for now Microsoft.FSharp.Control.Event type or Microsoft.FSharp.Control.IEvent<'Delegate,'Args> interface do not expose `GetInvocationList` method.
My current solution looks like this

``` F#
static let multicastFiled = typeof<Event<EventHandler, EventArgs>>.GetField("multicast", BindingFlags.NonPublic ||| BindingFlags.Instance)
let getDelegateFromEvent (event : Event<EventHandler, EventArgs>) = multicastFiled.GetValue event :?> EventHandler
```

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): XS

https://github.com/dotnet/fsharp/issues/10380#issuecomment-736584904

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [X] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
* [X] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by xperiandri** [comment]
@dsyme any considerations about this? Absence of this method blocks standard well-known event usage scenario

**Comment by dsyme** [comment]
Is there any reason not to simply redefine your own `Event` type which has this additional functionality?  There's nothing really special about `Event` in F# - the special type is `IEvent` (which we won't be changing).

**Comment by xperiandri** [comment]
Do you mean just copying this
https://github.com/dotnet/fsharp/blob/2708e585d92f2d56dc6252c86d2564ccd6a41d6e/src/fsharp/FSharp.Core/event.fs#L70
and adding the required member?

**Comment by xperiandri** [comment]
Maybe it is a better workaround than the reflection that I did. But it is the second time this year I implement an event and the second time I need that method...

**Comment by dsyme** [comment]
Well, I mean, I don't mind adding the member really.  It makes sense to do it.

So I will mark this as approved - and I'd be happy to see other additions to Event that make it more useful in practice.

**Comment by dsyme** [comment]
@xperiandri Do you think you could work on a PR an RFC? thanks

**Comment by jwosty** [comment]
Perhaps a better name would be GetSubscriberList. Or, maybe a HasSubscribers method would be a better alternative.

**Comment by xperiandri** [comment]
@dsyme yes, is there a tutorial on how to write RFC? Or should I just do it by looking at existing RFCs?

**Comment by xperiandri** [comment]
> Perhaps a better name would be GetSubscriberList. Or, maybe a HasSubscribers method would be a better alternative.

By the way, you are right. I actually needed `Has`Subscriber` more than the list itself.

**Comment by jwosty** [comment]
> 
> 
> @dsyme yes, is there a tutorial on how to write RFC? Or should I just do it by looking at existing RFCs?

There's no tutorial; however there is [a Readme](https://github.com/fsharp/fslang-design) that explains the process well. Looking at other RFCs for reference is a good idea, too.

> By the way, you are right. I actually needed HasSubscriber` more than the list itself.

Right -- returning the actual list feels like it's leaking too much internals to me. But @dsyme approved it either way. This could be a good discussion for the RFC thread once you create it.

**Comment by dsyme** [comment]
>  returning the actual list feels like it's leaking too much internals to me. 

Yes, for me too.  Would prefer HasSubscribers

**Comment by dsyme** [comment]
@xperiandri Would you like to implement this?

**Comment by xperiandri** [comment]
Yes, probably this or next month
